src/runtime/contrib/tensorrt/tensorrt_builder.cc

```diff
@@ -0,0 +1,421 @@
+              LOG(FATAL) << "TRT requires a constant input here.";
+            }
+          } else {
+            LOG(FATAL) << "TRT requires a constant input here.";
+          }
+        } else {
+          LOG(FATAL) << "TRT requires a constant input here.";
+        }
+      }
+    } else {
+      VisitExpr(call->args[i]);
+    }
+  }
+
+  // Get inputs.
+  for (size_t i = 0; i < call->args.size(); ++i) {
+    auto it = node_output_map_.find(call->args[i].operator->());
+    CHECK(it != node_output_map_.end()) << "Input was not found.";
+    for (auto out : it->second) {
+      params.inputs.push_back(out);
+    }
+  }
+  if (!converter->variable_input_count) {
+    CHECK_EQ(converter->input_types.size(), params.inputs.size())
+        << "Op expected a different number of inputs.";
+  }
+
+  // Convert op to TRT.
+  converter->Convert(&params);
+
+  // Get outputs.
+  node_output_map_[call] = {};
+  std::vector<TrtOpInput> outputs;
+  for (auto out : params.outputs) {
+    node_output_map_[call].push_back(TrtOpInput(out));
+  }
+}
+
+int TensorRTBuilder::TrackVarNode(const VarNode* node) {
+  // TODO(trevmorr): make more robust
+  const int trim_length = std::string("tensorrt_input").length();
+  int var_node_idx =
+      std::stoi(node->name_hint().substr(trim_length, std::string::npos));
+  return var_node_idx;
+}
+
+void TensorRTBuilder::CleanUp() {
+  network_->destroy();
+  builder_->destroy();
+  for (auto weight : trt_weights_) {
+    if (weight.type == nvinfer1::DataType::kFLOAT) {
+      delete[] static_cast<const float*>(weight.values);
+    } else {
+      delete[] static_cast<const uint16_t*>(weight.values);
+    }
+  }
+}
+
+void TransposeRSCKtoKCRS(const std::vector<int>& original_shape,
+                         const float* input_values, float* output_values) {
+  const int r = original_shape[0];
+  const int s = original_shape[1];
+  const int c = original_shape[2];
+  const int k = original_shape[3];
+  for (int x = 0; x < k; x++) {
+    for (int y = 0; y < c; y++) {
+      for (int z = 0; z < r; z++) {
+        for (int w = 0; w < s; w++) {
+          const int input_index = (x) + (y * k) + (z * s * c * k) + (w * c * k);
+          const int output_index =
+              (x * c * r * s) + (y * r * s) + (z * s) + (w);
+          output_values[output_index] = input_values[input_index];
+        }
+      }
+    }
+  }
+}
+
+void TransposeRSCKtoCKRS(const std::vector<int>& original_shape,
+                         const float* input_values, float* output_values) {
+  const int r = original_shape[0];
+  const int s = original_shape[1];
+  const int c = original_shape[2];
+  const int k = original_shape[3];
+  for (int x = 0; x < k; x++) {
+    for (int y = 0; y < c; y++) {
+      for (int z = 0; z < r; z++) {
+        for (int w = 0; w < s; w++) {
+          const int input_index = (x) + (y * k) + (z * s * c * k) + (w * c * k);
+          const int output_index =
+              (y * k * r * s) + (x * r * s) + (z * s) + (w);
```

Comment:
Looks like this is the only line which is different in `TransposeRSCKtoCKRS` in comparison to `TransposeRSCKtoKCRS`. Maybe define two functions to calculate output_index for CKRS and KCRS and use generic implementation + corresponding function to get one or another behavior?

Reply:
Thanks, I created `TransposeWeights4D` which takes input and output strides as arguments.